### PR TITLE
feat: allow any GitHub path on the `upload_files` instruction

### LINKED ISSUE #1700: unable to use arbitrary files from non-package sources
### Background & motivation

Today, Kurtosis only supports the mounting of files to containers when those files are part of a Kurtosis package. This is painful for when I need to mount files like Grafana dashboard templates to my package for rendering the correct dashboard.

My workarounds today is to store my files in a child/sub package and maintain it. Then, in my main package, import in the child/sub package (via `import_module()`)

### Desired behaviour

A Kurtosis package author should be able to specify any local file path or URL to fetch data from & turn into a files artifact to use for the services inside their enclaves, without needing to have the desired data to be part of a Kurtosis package.

### How important is this to you?

Painful; the lack of this feature makes using Kurtosis frictionful.

### What area of the product does this pertain to?

CLI: the Command Line Interface

### PR DESCRIPTION
**Pull request by leoporoli**
## Description:
allow any GitHub path on the `upload_files` instruction

## Is this change user facing?
YES

## References (if applicable):
Fix https://github.com/kurtosis-tech/kurtosis/issues/1700

### CODE REVIEW

File: api/golang/core/lib/shared_utils/parsed_git_url.go [hunk 1]
```diff
@@ -12,7 +12,7 @@ const (
 	GithubDomainPrefix = "github.com"
 	httpsSchema        = "https"
 	UrlPathSeparator   = "/"
-	//MinimumSubPathsForValidGitURL for a valid GitURl we need it to look like github.com/author/moduleName
+	//MinimumSubPathsForValidGitURL for a valid GitURl we need it to look like github.com/author/repositoryName
```
Comment (h4ck3rk3y) [inline api/golang/core/lib/shared_utils/parsed_git_url.go:15]: very thorough!

File: core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go [hunk 2]
```diff
@@ -82,24 +81,34 @@ func (provider *GitPackageContentProvider) GetKurtosisYaml(packageAbsolutePathOn
...
+	shouldOnlyAcceptsPackagePath := false
+	return provider.getOnDiskAbsolutePath(repositoryPathURL, shouldOnlyAcceptsPackagePath)
+}
+
+func (provider *GitPackageContentProvider) GetOnDiskAbsolutePackageFilePath(repositoryPathURL string) (string, *startosis_errors.InterpretationError) {
+	shouldOnlyAcceptsPackagePath := true
+	return provider.getOnDiskAbsolutePath(repositoryPathURL, shouldOnlyAcceptsPackagePath)
+}
+
+func (provider *GitPackageContentProvider) getOnDiskAbsolutePath(repositoryPathURL string, shouldOnlyAcceptsPackageFilePath bool) (string, *startosis_errors.InterpretationError) {
```
Comment (h4ck3rk3y) [inline core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go:102]: neat! :-)

File: core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go [hunk 3]
```diff
@@ -114,30 +123,31 @@ func (provider *GitPackageContentProvider) GetOnDiskAbsoluteFilePath(absoluteFil
...
+	if interpretationError := provider.atomicClone(parsedURL); interpretationError != nil {
 		return "", interpretationError
 	}
 
-	// check whether kurtosis yaml exists in the path
-	maybeKurtosisYamlPath, interpretationError := getKurtosisYamlPathForFileUrl(pathToFileOnDisk, provider.packagesDir)
-	if interpretationError != nil {
-		return "", startosis_errors.WrapWithInterpretationError(err, "Error occurred while verifying whether '%v' belongs to a Kurtosis package.", absoluteFileLocator)
-	}
+	if shouldOnlyAcceptsPackageFilePath {
```
Comment (h4ck3rk3y) [inline core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go]: i wonder if we can check if this is not true and eject immediately with `return pathTofileOnDisk, nil` and then do the checks if it is true

saves some indentation, is in line with ejecting early

File: core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go [hunk 4]
```diff
@@ -301,6 +311,9 @@ func (provider *GitPackageContentProvider) atomicClone(parsedURL *shared_utils.P
 		depth = depthAssumingBranchTagsCommitsAreSpecified
 	}
 
+	//TODO evaluate to use the GitHub client GetContents call instead, because we are cloning the entire repository with this approach
```
Comment (h4ck3rk3y) [inline core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go]: aren't we doing a shallow clone!?

File: core/server/api_container/server/startosis_engine/kurtosis_instruction/upload_files/upload_files.go [hunk 5]
```diff
@@ -116,7 +116,7 @@ func (builtin *UploadFilesCapabilities) Interpret(locatorOfModuleInWhichThisBuil
 		return nil, startosis_errors.WrapWithInterpretationError(interpretationErr, "Tried to convert locator '%v' into absolute locator but failed", src.GoString())
 	}
 
-	pathOnDisk, interpretationErr := builtin.packageContentProvider.GetOnDiskAbsoluteFilePath(absoluteLocator)
+	pathOnDisk, interpretationErr := builtin.packageContentProvider.GetOnDiskAbsolutePath(absoluteLocator)
```
Comment (h4ck3rk3y) [inline core/server/api_container/server/startosis_engine/kurtosis_instruction/upload_files/upload_files.go:119]: nice, so `read_file` and `import_module` go unaffected

File: core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider_test.go [hunk 6]
```diff
@@ -188,19 +189,55 @@ func TestGetAbsolutePathOnDisk_WorksForNonInMainBranchLocators(t *testing.T) {
...
 
 	absoluteFileLocator := "github.com/kurtosis-tech/sample-dependency-package@test-branch/main.star"
-	pathOnDisk, err := provider.GetOnDiskAbsoluteFilePath(absoluteFileLocator)
+	pathOnDisk, err := provider.getOnDiskAbsolutePath(absoluteFileLocator, true)
 
 	require.Nil(t, err, "This test depends on your internet working and the kurtosis-tech/datastore-army-package existing")
 	require.Equal(t, path.Join(packageDir, "kurtosis-tech", "sample-dependency-package", "main.star"), pathOnDisk)
 }
 
+func TestGetAbsolutePathOnDisk_GenericRepositoryDir(t *testing.T) {
```
Comment (h4ck3rk3y) [inline core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider_test.go:205]: Should we also add an e2e test for upload_files?

**Review by h4ck3rk3y — APPROVED** [review]

File: core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go (hunk 3 above)
Comment (leoporoli) [reply core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go]: great suggestion Keep doing it @h4ck3rk3y

File: core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go (hunk 4 above)
Comment (leoporoli) [reply core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider.go]: yes, we are doing a shallow clone, there is a mistake in the comment, I was referring to the `workspace` but I put repository. I will change it.
I'm referring for instance if we want just only one file from a repository's workspace we are cloning the entire repository's workspace, but I think that we could get only that one file by using the GitHubClient.GetContents() call

File: core/server/api_container/server/startosis_engine/kurtosis_instruction/upload_files/upload_files.go (hunk 5 above)
Comment (leoporoli) [reply core/server/api_container/server/startosis_engine/kurtosis_instruction/upload_files/upload_files.go:119]: yes

File: core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider_test.go (hunk 6 above)
Comment (leoporoli) [reply core/server/api_container/server/startosis_engine/startosis_packages/git_package_content_provider/git_package_content_provider_test.go:205]: it was added, thanks.

**Comment by barnabasbusa** [comment]
Does this also support random urls? 
e.g: `https://ephemery.dev/test-genesis.tar.gz`

**Comment by leoporoli** [comment]
@barnabasbusa sorry, it doesn't. It only supports "GitHub" URL's

**Pull request merged**